src/mc-fle2-payload-uev-common.c

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright 2023-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mc-fle2-payload-uev-common-private.h"
+#include "mongocrypt-private.h"
+
+bool
+_mc_FLE2UnindexedEncryptedValueCommon_parse (const _mongocrypt_buffer_t *buf,
+                                             uint8_t *fle_blob_subtype,
+                                             uint8_t *original_bson_type,
+                                             _mongocrypt_buffer_t *key_uuid,
+                                             _mongocrypt_buffer_t *ciphertext,
+                                             mongocrypt_status_t *status)
+{
+   BSON_ASSERT_PARAM (buf);
+   BSON_ASSERT_PARAM (fle_blob_subtype);
+   BSON_ASSERT_PARAM (original_bson_type);
+   BSON_ASSERT_PARAM (key_uuid);
+   BSON_ASSERT_PARAM (ciphertext);
+
+   uint32_t offset = 0;
```

Comment:
You can use the new src/mc-reader-private.h API to reduce all the copy/paste bounds checking. See mc_FLE2IndexedEqualityEncryptedValueTokens_init_from_buffer

Reply:
very nice! done.